tests/integration/helpers/cluster.py

```diff
@@ -4464,6 +4471,14 @@ def write_embedded_config(name, dest_dir, fix_log_level=False):
                     "- " + external_dir_abs_path + ":" + external_dir + "\n"
                 )
 
+        # The current implementation of `self.env_variables` is not exclusive. Meaning the variables
+        # are shared with all nodes within the same cluster, even if it is specified for a single node.
+        # In order not to break the existing tests, the `self.exclusive_env_variables` option was added as a workaround.
+        # IMHO, it would be better to make `self.env_variables` exclusive by defaultand remove the `self.exclusive_env_variables` option.
+        if self.exclusive_env_variables:
```

Comment:
Maybe better to put it into 'environment' element, so it overwrites whatever is set in `env_file`? That way we can have both:
- some cluster-generic envs set via `env_file` element (cluster argument `env_variables`)
- instance-specific envs set via `environment` element (cluster argument `instance_env_variables`)

Reply:
We don't want to override it, we need the variables that are currently being set in the root `.env` file.

Btw, what do you mean with "put it into environment element". You mean use the `environment: ` section of docker compose (i.e, modify template file) or make the `exclusive_env_variables` a map object instead of a boolean?

Reply:
The current implementation creates a new file, fills it in with the env variables set in global `.env` and adds the "exclusive" ones.

Implementing the former would have the same effect if I am not missing anything.

The latter (`instance_env_variables` as a map type) would allow the user to set both cluster and instance variables in a single `add_instance` call. I can make that happen, it should be fairly easy.

Reply:
I meant "put it into `environment:` element in service definition".

In that case "overriding" means that everything from the `env_file` is taken into account, but values from `environment:` take precedence.

On the second thought, since you merge the dictionaries, proposed implementation does approximately the same thing.